src/core/mac/mac.cpp

```diff
@@ -269,16 +269,25 @@ void Mac::PerformEnergyScan(void)
 
     SuccessOrExit(error = UpdateScanChannel());
 
-    if (!(otPlatRadioGetCaps(&GetInstance()) & OT_RADIO_CAPS_ENERGY_SCAN) || (mScanDuration == 0))
+    if (mScanDuration == 0)
+    {
+        while (true)
+        {
+            int8_t rssi;
+
+            RadioReceive(mScanChannel);
+            rssi = otPlatRadioGetRssi(&GetInstance());
+            ReportEnergyScanResult(rssi);
+            SuccessOrExit(error = UpdateScanChannel());
+        }
+    }
+
+    if ((otPlatRadioGetCaps(&GetInstance()) & OT_RADIO_CAPS_ENERGY_SCAN) == 0)
```

Comment:
nit: change this to `else if` to make clear that this only occurs when `mScanDuration != 0`?